src/txrequest.cpp

```diff
@@ -0,0 +1,241 @@
+        if (MakeCompleted(m_index.project<ByTxid>(it))) {
+            // Then actually delete the entry (unless it was already deleted by MakeCompleted).
+            Erase<ByPeer>(it);
+        }
+        it = it_next;
+    }
+}
+
+void TxRequestTracker::ReceivedInv(uint64_t peer, bool outbound, const uint256& txid, std::chrono::microseconds now)
+{
+    // Bail out if we already have a CANDIDATE_BEST entry for this (txid, peer) combination. The case where there is
+    // a non-CANDIDATE_BEST entry already will be caught by the uniqueness property of the ByPeer index
+    // automatically.
+    if (m_index.get<ByPeer>().count(EntryPeer{peer, true, txid})) return;
+
+    auto time = outbound ? now : now + m_inbound_delay;
+    auto ret = m_index.get<ByPeer>().emplace(txid, peer, !outbound, time, m_sequence);
+    if (ret.second) {
+        auto it = m_index.project<ByTxid>(ret.first);
+        ++m_peerinfo[peer].m_total;
+        ++m_sequence;
+        if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != txid) &&
+            (std::next(it) == m_index.get<ByTxid>().end() || std::next(it)->m_txid != txid)) {
+            // This is both the first and the last entry for a given txid; set its m_first.
+            ret.first->m_first = true;
+        }
+    }
+}
+
+void TxRequestTracker::RequestedTx(uint64_t peer, const uint256& txid, std::chrono::microseconds now)
+{
+    auto it = m_index.get<ByPeer>().find(EntryPeer{peer, true, txid});
+    // RequestedTx can only be called on CANDIDATE_BEST entries (this is implied by its condition that it can only be
+    // called on txids returned by GetRequestable (and only AlreadyHave and RequestedTx can be called in between,
+    // which preserve the state of other txids).
+    assert(it != m_index.get<ByPeer>().end());
+    assert(it->GetState() == State::CANDIDATE_BEST);
+    Modify<ByPeer>(it, [now,this](Entry& entry) {
+        entry.SetState(State::REQUESTED);
+        entry.m_time = now + m_timeout;
+    });
+}
+
+void TxRequestTracker::ReceivedResponse(uint64_t peer, const uint256& txid)
+{
+    // We need to search the ByPeer index for both (peer, false, txid) and (peer, true, txid).
+    auto it = m_index.get<ByPeer>().find(EntryPeer{peer, false, txid});
+    if (it == m_index.get<ByPeer>().end()) it = m_index.get<ByPeer>().find(EntryPeer{peer, true, txid});
+    if (it != m_index.get<ByPeer>().end()) MakeCompleted(m_index.project<ByTxid>(it));
+}
+
+size_t TxRequestTracker::CountInFlight(uint64_t peer) const
+{
+    auto it = m_peerinfo.find(peer);
+    if (it != m_peerinfo.end()) return it->second.m_requested;
+    return 0;
+}
+
+size_t TxRequestTracker::CountTracked(uint64_t peer) const
+{
+    auto it = m_peerinfo.find(peer);
+    if (it != m_peerinfo.end()) return it->second.m_total;
+    return 0;
+}
+
+std::vector<uint256> TxRequestTracker::GetRequestable(uint64_t peer, std::chrono::microseconds now)
+{
+    // Move time.
+    SetTimePoint(now);
```

Comment:
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

Could passing the `peer` to `SetTimePoint()` improves efficiency by not processing entries related to other peers?

Reply:
That's a good suggestion, but I don't think it would actually be desirable (but fortunately, it also wouldn't gain any efficiency). An earlier design of this code worked like that, but I decided to move off that.

Say we're calling `GetRequestable` for peer A. There is a txid T that was announced by peer A and peer B. For some reason, peer B is currently the `CANDIDATE_BEST` peer for T, but its exptime is in the past. That means that the (T,B) pair in reality should be `COMPLETED` now (but the data structure hasn't been updated to reflect that), and peer A is actually the best match. If `GetRequestable` for A would only invoke `SetTimePoint` for currently-`CANDIDATE_BEST` entries, T won't be returned, at it'll only be when B calls `GetRequestable` that its expiration happens. This is probably not terrible behavior, but makes it *much* harder to exactly specify the behavior, and perhaps it introduces a bias in the selection - it's hard to say.

Thankfully, it also doesn't gain anything. Every (txid,peer) pair only ever transitions once to `COMPLETED`, so certainly in bulk work there cannot be any difference. Furthermore, `SetTimePoint` only iterates over all expired entries in its first loop, as they're sorted by time. So no time is wasted processing unexpired entries.

A similar reasoning applies to the second loop.